src/solvers/elliptic/linearSolver/PCG_EIGEN.cpp

```diff
@@ -0,0 +1,339 @@
+          s = 1.0;
+          c = 1.0;
+          p = 0.0;
+    
+          // QR iteration using givens' rotation
+          for (i=m-1;i>=l;i--) {
+            f = s * e[i-1];
+            b = c * e[i-1];
+    
+            if (std::abs(f) >= std::abs(g)) {
+              c = g / f;
+              r = sqrt(c*c+1.0);
+              e[i] = f*r;
+              s = 1.0/r;
+              c = c * s;
+            } else {
+              s = f / g;
+              r = sqrt(s*s+1.0);
+              e[i] = g*r;
+              c =  1.0 / r;
+              s = s * c;
+            }
+    
+            g = d[i] - p;
+            r = (d[i-1]-g) * s + 2.0 * c * b;
+            p = s * r;
+            d[i] = g + p;
+            g = c*r - b;
+          } // 14
+    
+          d[l-1] = d[l-1] - p;
+          e[l-1] = g;
+          e[m-1] = 0.0;
+        } // m<=l
+
+      } // m_in loop, exit at m_in==n
+      while (iskip==0 && m_in<n-1); 
+
+    } // outer iteration
+    while (iskip==0 && iter<=30);
+
+  } // l loop
+
+  dmax = 0.0;
+  dmin = d[0];
+
+  for (i=1;i<=n;i++) {
+    dmax = std::abs(std::max(d[i-1],dmax));
+    dmin = std::abs(std::min(d[i-1],dmin));
+  }
+  if(platform->comm.mpiRank == 0 && verbose) {
+    for (i=1;i<=n;i++) {
+      printf("PCG Ritz values %d / %d   %.6e \n", i, n, d[i-1]);
+    }
+  }
+  free(d);
+  free(e);
+}
+
+
+static dfloat update(elliptic_t* elliptic,
+                     occa::memory &o_p, occa::memory &o_Ap, const dfloat alpha,
+                     occa::memory &o_x, occa::memory &o_r)
+{
+  mesh_t* mesh = elliptic->mesh;
+
+  const bool serial = platform->serial;
+
+  // r <= r - alpha*A*p
+  // dot(r,r)
+  elliptic->updatePCGKernel(mesh->Nlocal,
+                            elliptic->fieldOffset,
+                            elliptic->o_invDegree,
+                            o_Ap,
+                            alpha,
+                            o_r,
+                            elliptic->o_tmpNormr);
+
+  dfloat rdotr1 = 0;
+#ifdef ELLIPTIC_ENABLE_TIMER
+    //platform->timer.tic("dotp",1);
+#endif
+  if(serial) {
+    rdotr1 = *((dfloat *) elliptic->o_tmpNormr.ptr());
+  } else {
+    const dlong Nblock = (mesh->Nlocal + BLOCKSIZE - 1) / BLOCKSIZE;
+    elliptic->o_tmpNormr.copyTo(elliptic->tmpNormr, Nblock*sizeof(dfloat));
+    for(int n = 0; n < Nblock; ++n)
+      rdotr1 += elliptic->tmpNormr[n];
+  }
+
+  // x <= x + alpha*p
+  platform->linAlg->axpbyMany(
+    mesh->Nlocal,
+    elliptic->Nfields,
+    elliptic->fieldOffset,
+    alpha,
+    o_p,
+    1.0,
+    o_x);
+
+  MPI_Allreduce(MPI_IN_PLACE, &rdotr1, 1, MPI_DFLOAT, MPI_SUM, platform->comm.mpiComm);
+#ifdef ELLIPTIC_ENABLE_TIMER
+    //platform->timer.toc("dotp");
+#endif
+
+  platform->flopCounter->add(elliptic->name + " ellipticUpdatePC",
+                             elliptic->Nfields * static_cast<double>(mesh->Nlocal) * 6 + mesh->Nlocal);
+
+  return rdotr1;
+}
+
+int pcg_eigen(elliptic_t* elliptic, occa::memory &o_r, occa::memory &o_x,
```

Comment:
Most of this is a duplication from `pcg` -- could we instead query the `options` object in the existing `pcg` on whether-or-not to provide eigenvalue estimates?